### PR TITLE
Define vertex data conversions and validation

### PR DESCRIPTION
**Pull request by kvark**
Fixes some of #1880

***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/1908.html" title="Last updated on Jul 7, 2021, 6:02 PM UTC (08367fe)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1908/3516793...kvark:08367fe.html" title="Last updated on Jul 7, 2021, 6:02 PM UTC (08367fe)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -4867,8 +4869,9 @@ dictionary GPUFragmentState: GPUProgrammableStage {
                 - |colorState|.{{GPUColorTargetState/blend}}.{{GPUBlendState/alpha}}
                     must be a [=valid GPUBlendComponent=].
             - |colorState|.{{GPUColorTargetState/writeMask}} must be &lt; 16.
-            - |descriptor|.{{GPUProgrammableStage/module}} must contain an output variable that:
-                - is [=statically used=] by |descriptor|.{{GPUProgrammableStage/entryPoint}}, and
+            - |descriptor|.{{GPUProgrammableStage/entryPoint}} must have a [=pipeline output=] value:
```
Comment (alan-baker) [inline spec/index.bs:4872]: The reverse should also be checked. Each pipeline output must have location that matches a target colorState.

Comment (Kangz) [reply spec/index.bs:4872]: That seems to match what @austinEng found in #1900 although we can probably fix this in a follow-up PR.

File: spec/index.bs [hunk 2]
```diff
@@ -5170,10 +5173,18 @@ dictionary GPUVertexAttribute {
...
         - For every vertex attribute in the shader reflection of |vertexStage|.{{GPUProgrammableStage/module}}
-            that is know to be [=statically used=] by |vertexStage|.{{GPUProgrammableStage/entryPoint}},
+            that is a [=pipeline input=] of |vertexStage|.{{GPUProgrammableStage/entryPoint}},
             there is a corresponding |attrib| element of |descriptor|.{{GPUVertexBufferLayout/attributes}} for which
             all of the following are true:
-            - The shader format is |attrib|.{{GPUVertexAttribute/format}}.
+            - The shader format is compatible with |attrib|.{{GPUVertexAttribute/format}}'s [=vertex data type=]:
+                <dl class="switch">
+                    : unorm", "snorm", or "float"
+                    :: shader format has to `f32` or `vecN<f32>`.
```
Comment (Kangz) [inline spec/index.bs:5182]: nit: has to be, ditto below

File: spec/index.bs [hunk 3]
```diff
@@ -8713,16 +8724,30 @@ clip space positions for [[#primitive-clipping]], as well as other data for the
                         2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
                         as `vec2<f32>` in WGSL.
                     </div>
+                1. Adjust the |data| size to the shader type:
+                    - if both are scalar, or both are vectors of the same dimensionality, no adjustment is needed.
+                    - if |data| is vector but the shader type is scalar, then only the first component is extracted.
+                    - if both are vectors, and |data| has a h igher dimension, the extra components are dropped.
```
Comment (Kangz) [inline spec/index.bs:8730]:
```suggestion
                    - if both are vectors, and |data| has a higher dimension, the extra components are dropped.
```

**Review by Kangz — APPROVED** [review]

File: spec/index.bs (hunk 1 above)
Comment (kvark) [reply spec/index.bs:4873]: Yeah, let's focus on the vertex data in this PR

**Review by toji — APPROVED** [review]

**Pull request merged**